src/nagini_translation/analyzer.py

```diff
@@ -460,15 +460,53 @@ def visit_ClassDef(self, node: ast.ClassDef) -> None:
                 if not (isinstance(base, ast.Name) and
                         base.id in MYPY_SUPERCLASSES):
                     actual_bases.append(base)
+
+        if len(actual_bases) > 0:
+            cls.superclass = self.find_or_create_target_class(actual_bases[0])
+
+            # Process ADT defined by a class hierarchy (sum) and a NamedTuple (product)
+            if isinstance(cls.superclass, PythonClass) and cls.superclass.is_adt:
+                cls.is_adt = True
+                # Detect malformed ADTs
+                # ADT classes can have at most two superclasses
+                if len(actual_bases) > 2:
+                    raise InvalidProgramException(cls.node, 'malformed.adt',
+                            'malformed algebraic data type: superclasses can only ' +
+                            'be a class optionally followed by one NamedTuple')
+                # ADT classes should have no body
+                if len(node.body) != 1 or not isinstance(node.body[0], ast.Pass):
```

Comment:
This would not allow 
```
class MyADT(ADT):
    """ Explanation of what the ADT does in a docstring """
```
IMO it would be good if we could support this.

Reply:
That's right, I fixed that and added a test case to enforce such possibility now.

Reply:
Great, thanks!